### PR TITLE
Add Range overlap and set operations

### LINKED ISSUE #10148: Add `Range#clamp`
Range isn't `Comparable`, but it could very well define a `#clamp` method. The result would be the intersection of both ranges.

```cr
(-5..5).clamp(0..100)    # =>  0..5
(-5..5).clamp(-100..100) # => -5..5
(-5..5).clamp(..)        # => -5..5
(-5...5).clamp(-100...0) # => -5...0
```

This method could also be called `#intersect` or `#&`, but IMO `#clamp` fit's best because it closely resembles `Comparable#clamp`.

There are a few cases where the expected results may be up for discussion:

* Mixing exclusive and inclusive ranges (`(-5..5).clamp(-10...2)` and `(-5...5).clamp(-10..2)`) should probably have the same result as if both ranges were exclusive.
* If both ranges are disparate (`(-5..5).clamp(10..100)`), the result should probably be an empty range. This can be expressed as an exclusive range with identical begin and end (`x...x`). But what should `x` be? Probably best to use the begin of the clamped range (`-5...-5`).
* If the direction of the ranges is different (`-5..5.clamp(10..-10)`), we could try to sort things out, but I suppose it's probably better to just treat them as disparate and return an empty range.

### LINKED ISSUE #14487: Add a method to `Range` to detect overlap/intersection with another `Range`
The `Range` type doesn't currently tell you if it overlaps with another instance, but it's a useful thing to check for. For example, a scheduling app may need to detect whether a given time period conflicts with any existing meetings/events:

```crystal
existing_events.any?(&.overlaps?(range))
```

### LINKED ISSUE #14488: Union and intersection of `Range`s
Combined with #14487, it would be useful to have `Range` methods that, when given another range, would return a new range that contains the union or intersection of both ranges.

For example:

```crystal
a = 1..5
b = 3..7

a.union(b)        # => 1..7
a.intersection(b) # => 3..5
```

If there is no overlap between them, I'm not 100% sure what should happen but I feel like returning `nil` makes sense.

```crystal
a = 1..5
c = 10..20

a.union(c)        # => nil
a.intersection(c) # => nil
```

There are probably some other edge cases to work out, as well, such as:
- what happens when ranges are infinite in one direction?
  - should the `union` be infinite in that direction?
  - should the `intersection` use whichever one has a finite value?
- for ranges like `1...10` and `10..20` where there's nooverlap, but they _are_ adjacent:
  - should the union be `1..20` — the union of adjacent ranges providing continuity
  - should the intersection be `nil` (or whatever the non-overlap result is deemed to be)?
- should `(1...10).union(1..10)` be `1..10`?
- should `(1...10).intersection(1..10)` be `1...10`?
- any other edge cases?

### PR DESCRIPTION
**Pull request by mamantoha**
This PR continues the work from https://github.com/crystal-lang/crystal/pull/15106.
Compared to that earlier PR, this version covers the edge cases raised in review comments: open-ended ranges, generic endpoint types, endpoint exclusivity, and adjacent discrete ranges using `succ`/`pred`.

It also keeps the `nil` return behavior for cases where `union` cannot produce a single continuous range or `intersection` has no overlap.

Fixes https://github.com/crystal-lang/crystal/issues/14487
Fixes https://github.com/crystal-lang/crystal/issues/14488
Fixes https://github.com/crystal-lang/crystal/issues/10148

### CODE REVIEW

File: spec/std/range_spec.cr [hunk 1]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
...
+    (1...1).overlaps?(1..1).should be_false
+    (2..1).overlaps?(1..2).should be_false
+    (2...1).overlaps?(1..2).should be_false
+    (1..2).overlaps?(2...2).should be_false
+
+    (..5).overlaps?(5..).should be_true
+    (...5).overlaps?(5..).should be_false
+    (..5).overlaps?(6..).should be_false
+    (..).overlaps?(1..2).should be_true
+    (..).overlaps?(2...2).should be_false
```
Comment (straight-shoota) [inline spec/std/range_spec.cr:126]: **suggestion:** Endless ranges have some nasty edge cases. We should cover more test cases for overlapping endless ranges.

I'm making some suggestions here and in following comments, but I'm probably missing some edge cases as well.

```suggestion
    (..5).overlaps?(5..).should be_true
    (...5).overlaps?(5..).should be_false
    (..5).overlaps?(6..).should be_false
    (..5).overlaps?(4..).should be_true
    (...5).overlaps?(4..).should be_true
    (..5).overlaps?(..4).should be_true
    (...5).overlaps?(..4).should be_true
    (..5).overlaps?(..6).should be_true
    (...5).overlaps?(..6).should be_true
    (5..).overlaps?(4..).should be_true
    (5..).overlaps?(6..).should be_true
    (..).overlaps?(1..2).should be_true
    (..).overlaps?(2...2).should be_false
```

File: spec/std/range_spec.cr [hunk 2]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
...
+    (1...6).union(6..10).should eq(1..10)
+
+    (1..5).union(10..20).should be_nil
+    (1...1).union(1..2).should eq(1..2)
+    (1..2).union(2...2).should eq(1..2)
+
+    (..5).union(5..).should eq(nil..nil)
+    (...5).union(5..).should eq(nil..nil)
+    (..5).union(6..).should eq(nil..nil)
+    (..5).union(7..).should be_nil
```
Comment (straight-shoota) [inline spec/std/range_spec.cr]: **suggestion:** ditto

```suggestion
    (..5).union(5..).should eq(nil..nil)
    (...5).union(5..).should eq(nil..nil)
    (..5).union(6..).should eq(nil..nil)
    (...5).union(6..).should be_nil
    (..5).union(7..).should be_nil
    (..5).union(..6).should eq(..6)
    (..5).union(4..).should eq(..)
    (...5).union(4..).should eq(..)
    (..).union(2..3).should eq(..)
    (...).union(2..3).should eq(...)
    (..).union(2...3).should eq(..)
```

File: spec/std/range_spec.cr [hunk 3]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
     (1...5).includes?(5).should be_false
   end
 
+  it "overlaps?" do
```
Comment (straight-shoota) [inline spec/std/range_spec.cr]:
```suggestion
  it "#overlaps?" do
```

File: spec/std/range_spec.cr [hunk 4]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
...
+    t3 = Time.local(2024, 10, 10)
+    t4 = Time.local(2024, 10, 15)
+    (t1..t3).overlaps?(t2..t4).should be_true
+    (t1..t2).overlaps?(t3..t4).should be_false
+
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(3)..RangeSpecIntWrapper.new(7)).should be_true
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(6)..RangeSpecIntWrapper.new(7)).should be_false
+  end
+
+  it "unions ranges" do
```
Comment (straight-shoota) [inline spec/std/range_spec.cr]:
```suggestion
  it "#unions" do
```

File: spec/std/range_spec.cr [hunk 5]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
...
+
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).union(RangeSpecIntWrapper.new(3)..RangeSpecIntWrapper.new(7)).should eq(RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(7))
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).union(RangeSpecIntWrapper.new(6)..RangeSpecIntWrapper.new(7)).should eq(RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(7))
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).union(RangeSpecIntWrapper.new(7)..RangeSpecIntWrapper.new(8)).should be_nil
+
+    (RangeSpecPredWrapper.new(1)..RangeSpecPredWrapper.new(5)).union(RangeSpecPredWrapper.new(6)..RangeSpecPredWrapper.new(7)).should eq(RangeSpecPredWrapper.new(1)..RangeSpecPredWrapper.new(7))
+    (RangeSpecPredWrapper.new(1)..RangeSpecPredWrapper.new(5)).union(RangeSpecPredWrapper.new(7)..RangeSpecPredWrapper.new(8)).should be_nil
+  end
+
+  it "intersects ranges" do
```
Comment (straight-shoota) [inline spec/std/range_spec.cr]:
```suggestion
  it "#intersection" do
```

File: spec/std/range_spec.cr [hunk 6]
```diff
@@ -79,6 +96,147 @@ describe "Range" do
...
+    (2..10).intersection(0..8).should eq(2..8)
+    (1..5).intersection(5..10).should eq(5..5)
+
+    (1..5).intersection(10..20).should be_nil
+    (1...10).intersection(10..11).should be_nil
+    (1...1).intersection(1..2).should be_nil
+
+    (..5).intersection(5..).should eq(5..5)
+    (...5).intersection(5..).should be_nil
+    (..5).intersection(3..).should eq(3..5)
```
Comment (straight-shoota) [inline spec/std/range_spec.cr:224]:
```suggestion
    (..5).intersection(5..).should eq(5..5)
    (...5).intersection(5..).should be_nil
    (..5).intersection(3..).should eq(3..5)
    (..5).intersection(..3).should eq(..3)
    (..5).intersection(-5..).should eq(-5..5)
    (...5).intersection(-5..).should eq(-5...5)
    (..5).intersection(-5...).should eq(-5..5)
    (..5).intersection(2..7).should eq(2..5)
    (..5).intersection(2...7).should eq(2..5)
    (...5).intersection(2...7).should eq(2...5)
    (..5).intersection(2..4).should eq(2..4)
    (...5).intersection(2..4).should eq(2..4)
    (..5).intersection(2...4).should eq(2...4)
    (..).intersection(2..3).should eq(2..3)
    (..).intersection(2...3).should eq(2...3)
```

File: spec/std/range_spec.cr [hunk 7]
```diff
@@ -79,6 +96,174 @@ describe "Range" do
...
+    t3 = Time.local(2024, 10, 10)
+    t4 = Time.local(2024, 10, 15)
+    (t1..t3).overlaps?(t2..t4).should be_true
+    (t1..t2).overlaps?(t3..t4).should be_false
+
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(3)..RangeSpecIntWrapper.new(7)).should be_true
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(6)..RangeSpecIntWrapper.new(7)).should be_false
+  end
+
+  it "#unions" do
```
Comment (jgaskins) [inline spec/std/range_spec.cr]: Typo in the name of the method under test

```suggestion
  it "#union" do
```

File: spec/std/range_spec.cr [hunk 8]
```diff
@@ -79,6 +96,174 @@ describe "Range" do
...
+    (6..10).union(1..5).should eq(1..10)
+    (1...6).union(6..10).should eq(1..10)
+
+    (1..5).union(10..20).should be_nil
+    (1...1).union(1..2).should eq(1..2)
+    (1..2).union(2...2).should eq(1..2)
+
+    (..5).union(5..).should eq(nil..nil)
+    (...5).union(5..).should eq(nil..nil)
+    (..5).union(6..).should eq(nil..nil)
```
Comment (jgaskins) [inline spec/std/range_spec.cr]: To match the style of the ones below

```suggestion
    (..5).union(5..).should eq(..)
    (...5).union(5..).should eq(..)
    (..5).union(6..).should eq(..)
```

File: src/range.cr [hunk 9]
```diff
@@ -319,6 +434,34 @@ struct Range(B, E)
...
+        @exclusive ? begin_value >= end_value : begin_value > end_value
+      else
+        false
+      end
+    else
+      false
+    end
+  end
+
+  private def adjacent_to?(other : Range) : Bool
```
Comment (jgaskins) [inline src/range.cr:444]: How bad an idea would it be to make this method public?

File: src/range.cr [hunk 10]
```diff
@@ -319,6 +434,34 @@ struct Range(B, E)
...
+  protected def empty_without_iterating? : Bool
+    if end_value = @end
+      if begin_value = @begin
+        @exclusive ? begin_value >= end_value : begin_value > end_value
+      else
+        false
+      end
+    else
+      false
+    end
```
Comment (jgaskins) [inline src/range.cr]: Nil checks are great candidates for guard clauses here

```suggestion
    return false unless end_value = @end
    return false unless begin_value = @begin
    
    @exclusive ? begin_value >= end_value : begin_value > end_value
```

File: src/range.cr [hunk 11]
```diff
@@ -319,6 +434,34 @@ struct Range(B, E)
...
+    return true if end_value && other_begin && end_value == other_begin
+    return true if end_value && other_begin && end_value.responds_to?(:succ) && end_value.succ == other_begin
+    return true if end_value && other_begin && other_begin.responds_to?(:pred) && other_begin.pred == end_value
+
+    other_end = other.end
+    begin_value = @begin
+
+    return true if other_end && begin_value && other_end == begin_value
+    return true if other_end && begin_value && other_end.responds_to?(:succ) && other_end.succ == begin_value
+    !!(other_end && begin_value && begin_value.responds_to?(:pred) && begin_value.pred == other_end)
```
Comment (jgaskins) [inline src/range.cr]: I noticed the extended test cases that @straight-shoota recommended found a bug in this method.

```suggestion
    end_value = @end
    other_begin = other.begin

    return true if end_value && other_begin && end_value == other_begin
    unless excludes_end?
      return true if end_value && other_begin && end_value.responds_to?(:succ) && end_value.succ == other_begin
      return true if end_value && other_begin && other_begin.responds_to?(:pred) && other_begin.pred == end_value
    end

    other_end = other.end
    begin_value = @begin

    return true if other_end && begin_value && other_end == begin_value
    unless other.excludes_end?
      return true if other_end && begin_value && other_end.responds_to?(:succ) && other_end.succ == begin_value
      return true if other_end && begin_value && begin_value.responds_to?(:pred) && begin_value.pred == other_end
    end

    false
```

GitHub's suggested patches don't always make it easy to discern what changed, but it's this:

```diff
diff --git a/src/range.cr b/src/range.cr
index b0657a2a8..7e8ad2312 100644
--- a/src/range.cr
+++ b/src/range.cr
@@ -451,15 +446,21 @@ struct Range(B, E)
     other_begin = other.begin

     return true if end_value && other_begin && end_value == other_begin
+    unless excludes_end?
       return true if end_value && other_begin && end_value.responds_to?(:succ) && end_value.succ == other_begin
       return true if end_value && other_begin && other_begin.responds_to?(:pred) && other_begin.pred == end_value
+    end

     other_end = other.end
     begin_value = @begin

     return true if other_end && begin_value && other_end == begin_value
+    unless other.excludes_end?
       return true if other_end && begin_value && other_end.responds_to?(:succ) && other_end.succ == begin_value
-    !!(other_end && begin_value && begin_value.responds_to?(:pred) && begin_value.pred == other_end)
+      return true if other_end && begin_value && begin_value.responds_to?(:pred) && begin_value.pred == other_end
+    end
+
+    false
   end
```

**Review by jgaskins — COMMENTED** [review]
I really like the simplicity here. A couple minor notes and a fix for the failing spec, but this is really good. I ran a simple benchmark and both `union` and `overlaps?` execute in 1ns on my machine.

File: src/range.cr [hunk 12]
```diff
@@ -319,6 +434,35 @@ struct Range(B, E)
...
+
+  private def adjacent_to?(other : Range) : Bool
+    end_value = @end
+    other_begin = other.begin
+
+    return true if end_value && other_begin && end_value == other_begin
+    unless excludes_end?
+      return true if end_value && other_begin && end_value.responds_to?(:succ) && end_value.succ == other_begin
+      return true if end_value && other_begin && other_begin.responds_to?(:pred) && other_begin.pred == end_value
+    end
```
Comment (straight-shoota) [inline src/range.cr]: **polish:** We can extract the common part of the conditions:
```suggestion
    if end_value && other_begin
      return true if end_value == other_begin
      unless excludes_end?
        return true if end_value.responds_to?(:succ) && end_value.succ == other_begin
        return true if other_begin.responds_to?(:pred) && other_begin.pred == end_value
      end
    end
```

File: src/range.cr [hunk 13]
```diff
@@ -319,6 +434,35 @@ struct Range(B, E)
...
+
+    other_end = other.end
+    begin_value = @begin
+
+    return true if other_end && begin_value && other_end == begin_value
+    unless other.excludes_end?
+      return true if other_end && begin_value && other_end.responds_to?(:succ) && other_end.succ == begin_value
+      return true if other_end && begin_value && begin_value.responds_to?(:pred) && begin_value.pred == other_end
+    end
```
Comment (straight-shoota) [inline src/range.cr]: ditto


```suggestion
    if other_end && begin_value
      return true if other_end == begin_value
      unless other.excludes_end?
        return true if other_end.responds_to?(:succ) && other_end.succ == begin_value
        return true if begin_value.responds_to?(:pred) && begin_value.pred == other_end
      end
    end
```

File: spec/std/range_spec.cr [hunk 14]
```diff
@@ -79,6 +96,174 @@ describe "Range" do
...
+    t3 = Time.local(2024, 10, 10)
+    t4 = Time.local(2024, 10, 15)
+    (t1..t3).overlaps?(t2..t4).should be_true
+    (t1..t2).overlaps?(t3..t4).should be_false
+
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(3)..RangeSpecIntWrapper.new(7)).should be_true
+    (RangeSpecIntWrapper.new(1)..RangeSpecIntWrapper.new(5)).overlaps?(RangeSpecIntWrapper.new(6)..RangeSpecIntWrapper.new(7)).should be_false
+  end
+
+  it "#union" do
```
Comment (straight-shoota) [inline spec/std/range_spec.cr:145]: **issue:** I believe we're missing an explicit example for adjacent numbers in floating point notation:
```cr
(..3.0).union(4.0..).should be_nil
```
It's perhaps a bit surprising that only the type of the lower value matters:
```cr
(..3).union(4.0..) # => ..
```

And we could add some more examples with mixed types and ensuring the typed result is as expected:
```cr
(1..5).union(3.0..7.0) # => 1..7.0
(1..5).union(1.0...7)  # => 1...7
(1.0..5).union(1...7)  # => 1.0...7
```

Comment (mamantoha) [reply spec/std/range_spec.cr:145]: done

File: spec/std/range_spec.cr [hunk 15]
```diff
@@ -79,6 +96,182 @@ describe "Range" do
...
+    typeof((1..5).intersection(10..20)).should eq(Range(Int32, Int32)?)
+
+    typeof((1..5).union(3_i64..7_i64)).should eq(Range(Int32 | Int64, Int32 | Int64)?)
+    typeof((1..5).intersection(3_i64..7_i64)).should eq(Range(Int32 | Int64, Int32 | Int64)?)
+    typeof((1..5).union(3.0..7.0)).should eq(Range(Int32 | Float64, Int32 | Float64)?)
+    typeof((1..5).union(1.0...7)).should eq(Range(Int32 | Float64, Int32)?)
+    typeof((1.0..5).union(1...7)).should eq(Range(Int32 | Float64, Int32)?)
+
+    typeof((..5).union(3..)).should eq(Range(Int32 | Nil, Int32 | Nil)?)
+    typeof((..5).intersection(3..)).should eq(Range(Int32 | Nil, Int32 | Nil)?)
```
Comment (straight-shoota) [inline spec/std/range_spec.cr:272]: **thought:** `typeof` gives a static answer and primarily verifies that the return type restriction is correct.
When ranges with different types are involved, it might be more interesting to look at the _actual_ result type, for example that `(1.0..5).union(1...7).is_a(Range(Float64, Int32)`, because `eq 1.0..7` doesn't cover this type strictness (`eq 1..7` would succeed as well).
Maybe we could add a couple of these in cases where the result might be ambiguous? Or we skip it for now and wait for #13389 to resolve this implicitly.

**Comment by Sija** [comment]
Would be great to have it included in the Crystal v1.22.0 release... 🤞
